plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java

```diff
@@ -160,22 +168,50 @@ public JdbcTableHandle getTableHandle(ConnectorSession session, SchemaTableName
 
             newDomain = TupleDomain.withColumnDomains(supported);
             remainingFilter = TupleDomain.withColumnDomains(unsupported);
+
+            if (isComplexExpressionPushdown(session)) {
+                List<String> newExpressions = new ArrayList<>();
+                List<ConnectorExpression> remainingExpressions = new ArrayList<>();
+                for (ConnectorExpression expression : extractConjuncts(constraint.getExpression())) {
```

Comment:
This still looks coupled with predicates in WHERE, it won't fire for the join conditions?

As we work on more pushdown I'm seeing some duplication of work between the predicate handling and the join condition handling.

Is it possible to extract all "expressions" from a query and pass them through `JdbcClient#convertPredicate`?
Is there something special about join conditions that I'm missing?

Reply:
> This still looks coupled with predicates in WHERE, it won't fire for the join conditions?

Correct. We're within `applyFilter` function, so it's specific to WHERE, and doesn't apply to join conditions.

> As we work on more pushdown I'm seeing some duplication of work between the predicate handling and the join condition handling.

Join conditions aren't expressed as `ConnectorExpression` yet.
I don't see a future-compatibility problem when they become, except that we will need to support table aliases to distinguish between lhs and rhs columns.

> Is it possible to extract all "expressions" from a query and pass them through JdbcClient#convertPredicate?
Is there something special about join conditions that I'm missing?

Like combining `PushPredicateIntoTableScan` and `PushJoinIntoTableScan`?
It's mostly not needed, as the rules are applied iteratively. It may be needed for the case where Filter isn't getting pushed down but the Join would. However, this can be solved without merging the rules, if we decide this is something to solve.

Reply:
> Like combining PushPredicateIntoTableScan and PushJoinIntoTableScan?

Not really. What I mean is that instead of having different methods in JdbcClient which optimise the join and filters we can have both re-use `convertPredicate`.
e.g. if we wanted to support `OR` in join pushdown we'd end up with a copy of the code from `applyFilter`.

But yes agreed that it'll be easier to see what the solution should look like once we have a concrete usecase.

Reply:
The `JdbcClient` may have two methods and reuse `ConnectorExpressionRewriter`.
or it can have one method (the `convertPredicate`), with changed signature.
at this point it's unclear how `PushJoinIntoTableScan` will pass `ConnectorExpression` and how it will discern between `ColumnHandle`-s coming from left and right side of the join, so i don't know how to future-proof this code fully.